wiring/src/spark_wiring_ble.cpp

```diff
@@ -0,0 +1,1756 @@
+    return observerProxy_->scan(callback, timeout);
+}
+
+int BleLocalDevice::scan(BleScanResult* results, size_t resultCount) {
+    return observerProxy_->scan(results, resultCount);
+}
+
+int BleLocalDevice::scan(BleScanResult* results, size_t resultCount, uint16_t timeout) {
+    return observerProxy_->scan(results, resultCount, timeout);
+}
+
+int BleLocalDevice::scan(BleScanResult* results, size_t resultCount, const BleScanParams& params) {
+    return observerProxy_->scan(results, resultCount, params);
+}
+
+int BleLocalDevice::stopScanning(void) {
+    return observerProxy_->stopScanning();
+}
+
+int BleLocalDevice::setPPCP(uint16_t minInterval, uint16_t maxInterval, uint16_t latency, uint16_t timeout) {
+    return peripheralProxy_->setPPCP(minInterval, maxInterval, latency, timeout);
+}
+
+int BleLocalDevice::disconnect(void) {
+    return peripheralProxy_->disconnect();
+}
+
+bool BleLocalDevice::connected(void) const {
+    return (peripheralProxy_->connected() || centralProxy_->connected());
+}
+
+BlePeerDevice BleLocalDevice::connect(const BleAddress& addr, uint16_t interval, uint16_t latency, uint16_t timeout) {
+    hal_ble_conn_params_t connParams;
+    connParams.min_conn_interval = interval;
+    connParams.max_conn_interval = interval;
+    connParams.slave_latency = latency;
+    connParams.conn_sup_timeout = timeout;
+    ble_gap_set_ppcp(&connParams, nullptr);
+    return connect(addr);
+}
+
+BlePeerDevice BleLocalDevice::connect(const BleAddress& addr) {
+    BlePeerDevice peer = centralProxy_->connect(addr);
+    // Automatically discover services and characteristics.
+    if (peer.connected()) {
+        gattcProxy_->discoverAllServices(peer);
+    }
+    return peer;
+}
+
+int BleLocalDevice::disconnect(const BlePeerDevice& peripheral) {
+    return centralProxy_->disconnect(peripheral);
+}
+
+int BleLocalDevice::addCharacteristic(BleCharacteristic& characteristic) {
+    return gattsProxy_->addCharacteristic(characteristic);
+}
+
+int BleLocalDevice::addCharacteristic(const char* desc, BleCharacteristicProperties properties, OnDataReceivedCallback callback) {
+    BleCharacteristic characteristic(desc, properties, callback);
+    return gattsProxy_->addCharacteristic(characteristic);
+}
+
+BlePeerDevice* BleLocalDevice::findPeerDevice(BleConnectionHandle connHandle) {
+    for (size_t i = 0; i < peripheralProxy_->centralCount(); i++) {
+        BlePeerDevice* peer = &peripheralProxy_->centrals[i];
+        if (peer != nullptr) {
+            if (peer->connHandle == connHandle) {
+                return peer;
+            }
+        }
+    }
+    for (size_t i = 0; i < centralProxy_->peripheralCount(); i++) {
+        BlePeerDevice* peer = &centralProxy_->peripherals[i];
+        if (peer != nullptr) {
+            if (peer->connHandle == connHandle) {
+                return peer;
+            }
+        }
+    }
+    return nullptr;
+}
+
+void BleLocalDevice::onBleEvents(const hal_ble_evts_t *event, void* context) {
```

Comment:
All of these events come from a non-application thread. There should be locking primitives in place to provide thread safety between the application thread calling BLE functions and anything modified by the incoming events.

Reply:
^^ This is still an issue though.The use pattern should be to lock the wiring BLE lock only when accessing/modifying the data members both here and in the wiring methods. It should be released when calling into BLE HAL methods to avoid deadlocking.